craft_parts/constraints.py

```diff
@@ -43,3 +54,5 @@ def _validate_relative_path_str(path: str) -> str:
     # The field here is used to provide information in the JSON schema and IDEs.
     Field(description="relative path", min_length=1, pattern=re.compile(r"^[^\/].*")),
 ]
+
+UniqueList = Annotated[list[T], AfterValidator(_validate_list_is_unique)]
```

Comment:
It's worth considering if something like `craft_parts/types.py` wouldn't be a better place for this.

Reply:
I didn't want to call this file `types` because I wanted to make it obvious that the only types that should belong in here are constrained types for pydantic.